temporalio/worker/_tuning.py

```diff
@@ -0,0 +1,185 @@
+    Defaults to 0 for workflows and 50ms for activities.
+    
+    This value matters because how many resources a task will use cannot be determined ahead of time, and thus the
+    system should wait to see how much resources are used before issuing more slots."""
+
+
+@dataclass(frozen=True)
+class ResourceBasedSlotSupplier:
+    """A slot supplier that will dynamically adjust the number of slots based on resource usage.
+
+    .. warning::
+        The resource based tuner is currently experimental.
+    """
+
+    slot_options: ResourceBasedSlotOptions
+    tuner_options: ResourceBasedTunerOptions
+    """Options for the tuner that will be used to adjust the number of slots. When used with a
+    :py:class:`CompositeTuner`, all resource-based slot suppliers must use the same tuner options."""
+
+
+SlotSupplier: TypeAlias = Union[FixedSizeSlotSupplier, ResourceBasedSlotSupplier]
+
+
+def _to_bridge_slot_supplier(
+    slot_supplier: SlotSupplier, kind: Literal["workflow", "activity", "local_activity"]
+) -> temporalio.bridge.worker.SlotSupplier:
+    if isinstance(slot_supplier, FixedSizeSlotSupplier):
+        return temporalio.bridge.worker.FixedSizeSlotSupplier(slot_supplier.num_slots)
+    elif isinstance(slot_supplier, ResourceBasedSlotSupplier):
+        min_slots = 5 if kind == "workflow" else 1
+        max_slots = 500
+        ramp_throttle = (
+            timedelta(seconds=0) if kind == "workflow" else timedelta(milliseconds=50)
+        )
+        if slot_supplier.slot_options.minimum_slots is not None:
+            min_slots = slot_supplier.slot_options.minimum_slots
+        if slot_supplier.slot_options.maximum_slots is not None:
+            max_slots = slot_supplier.slot_options.maximum_slots
+        if slot_supplier.slot_options.ramp_throttle is not None:
+            ramp_throttle = slot_supplier.slot_options.ramp_throttle
+        return temporalio.bridge.worker.ResourceBasedSlotSupplier(
+            min_slots,
+            max_slots,
+            int(ramp_throttle / timedelta(milliseconds=1)),
+            temporalio.bridge.worker.ResourceBasedTunerOptions(
+                slot_supplier.tuner_options.target_memory_usage,
+                slot_supplier.tuner_options.target_cpu_usage,
+            ),
+        )
+    else:
+        raise TypeError(f"Unknown slot supplier type: {slot_supplier}")
+
+
+class WorkerTuner(ABC):
+    """WorkerTuners allow for the dynamic customization of some aspects of worker configuration"""
+
+    @abstractmethod
+    def _get_workflow_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _get_local_activity_task_slot_supplier(self) -> SlotSupplier:
+        raise NotImplementedError
+
+
+class ResourceBasedTuner(WorkerTuner):
```

Comment:
I am not sure this should exist now that I think about it. I think this can be on worker tuner:

```
class WorkerTuner(ABC):
    @staticmethod
    def create_resource_based(
        *,
        target_memory_usage: float,
        target_cpu_usage: float,
        workflow_slot_config: Optional[ResourceBasedSlotConfig] = None,
        activity_slot_config: Optional[ResourceBasedSlotConfig] = None,
        local_activity_slot_config: Optional[ResourceBasedSlotConfig] = None,
    ) -> WorkerTuner:
        # Create composite tuner
```

I have no strong opinion on those trailing 3 kwarg names (e.g. can be `workflow_config:`)

Reply:
K, I can go with that